### ISSUE TITLE
STORY-AI-001: Learning from Play (Reinforcement Learning)

### ISSUE THREAD

**Issue by Hulupeep** [Hulupeep/mbot_ruvector#86]
## Description

Basic reinforcement learning system where robot learns from game outcomes and user feedback to improve behavior.

## DOD Criticality
- [ ] **Future** - Can release without

## Contract References
- **Feature Contracts:** I-AI-001, I-AI-002, I-AI-003, I-AI-004
- **Journey Contract:** J-AI-LEARNING

## Invariants Referenced
- **I-AI-001**: Q-learning must converge within 1000 episodes for simple games
- **I-AI-002**: Learned policy must persist across sessions (model serialization)
- **I-AI-003**: Learning rate decay prevents oscillation after convergence
- **I-AI-004**: Exploration rate (epsilon) must decrease over time (epsilon-greedy)

### Scenario: Learn from Tic-Tac-Toe
```gherkin
Given robot plays 100 games of tic-tac-toe
And learning is enabled
When I check win rate after 100 games
Then win rate improves from 40% to 70%+
And robot adapts strategy
And learned policy persists after restart
```

### Scenario: Learn from User Feedback
```gherkin
Given robot performs behavior "cautious exploration"
When user rates behavior as "good" (thumbs up)
Then robot reinforces that behavior
And increases probability of similar actions by 10%
When user rates behavior as "bad" (thumbs down)
Then robot reduces that behavior
And decreases probability by 10%
```

### Scenario: Multi-Game Learning
```gherkin
Given robot learned tic-tac-toe strategy
When robot plays connect-four
Then learning transfers basic strategy
And robot learns game-specific tactics
And maintains separate Q-tables per game
```

### Scenario: Policy Persistence
```gherkin
Given robot learned strategy over 500 games
When robot restarts
Then learned policy loads from storage
And robot maintains 70%+ win rate
And no re-training required
```

## Acceptance Criteria Checklist
- [ ] Q-learning algorithm implemented
- [ ] Reward function design for tic-tac-toe
- [ ] Reward function design for connect-four
- [ ] User feedback integration (thumbs up/down)
- [ ] Policy updates based on feedback
- [ ] Model persistence (save/load Q-table)
- [ ] Learning visualization (win rate over time)
- [ ] Epsilon-greedy exploration
- [ ] Learning rate decay
- [ ] Convergence detection
- [ ] Multi-game support (separate Q-tables)

## Data Contract

```typescript
interface ReinforcementLearner {
  // Learning
  learn(state: State, action: Action, reward: number, nextState: State): void;
  selectAction(state: State, explore: boolean): Action;
  updatePolicy(feedback: UserFeedback): void;

  // Persistence
  savePolicy(gameType: string): Promise<void>;
  loadPolicy(gameType: string): Promise<Policy>;

  // Monitoring
  getMetrics(): LearningMetrics;
  resetLearning(gameType: string): void;
}

interface State {
  gameType: string;
  boardState: any;              // Game-specific
  features: number[];           // Extracted features
}

interface Action {
  type: string;                 // e.g., 'place_marker'
  params: any;                  // e.g., {row: 1, col: 2}
}

interface Policy {
  qTable: Map<string, Map<string, number>>;  // state -> action -> Q-value
  gameType: string;
  episodeCount: number;
  learningRate: number;
  discountFactor: number;
  epsilon: number;              // Exploration rate
  createdAt: number;
  updatedAt: number;
}

interface UserFeedback {
  behaviorId: string;
  rating: 'good' | 'bad' | 'neutral';
  timestamp: number;
}

interface LearningMetrics {
  episodeCount: number;
  averageReward: number;
  winRate: number;              // Last 100 games
  lossRate: number;
  drawRate: number;
  epsilonCurrent: number;       // Current exploration rate
  learningRateCurrent: number;
  convergenceScore: number;     // 0-1, 1 = converged
}

interface LearningConfig {
  learningRate: number;         // Alpha (default: 0.1)
  discountFactor: number;       // Gamma (default: 0.9)
  epsilonStart: number;         // Initial exploration (default: 1.0)
  epsilonEnd: number;           // Final exploration (default: 0.1)
  epsilonDecay: number;         // Decay rate (default: 0.995)
  maxEpisodes: number;          // Max training episodes
}

// Reward functions
interface RewardFunction {
  gameType: string;

  // Game outcome rewards
  winReward: number;            // Default: +1.0
  lossReward: number;           // Default: -1.0
  drawReward: number;           // Default: 0.0

  // Intermediate rewards
  goodMoveReward: number;       // Default: +0.1
  badMoveReward: number;        // Default: -0.1

  // User feedback rewards
  userGoodReward: number;       // Default: +0.5
  userBadReward: number;        // Default: -0.5
}
```

## data-testid Coverage

| Element | data-testid | Purpose |
|---------|-------------|---------|
| Learning toggle | `learning-toggle` | Enable/disable learning |
| Learning dashboard | `learning-dashboard` | Shows metrics |
| Win rate chart | `win-rate-chart` | Win rate over episodes |
| Episode counter | `episode-counter` | Total episodes trained |
| Epsilon display | `epsilon-display` | Current exploration rate |
| Reward display | `reward-display` | Average reward |
| Feedback buttons | `feedback-{good/bad}` | User feedback |
| Reset learning button | `reset-learning-btn` | Clear learned policy |
| Save policy button | `save-policy-btn` | Save Q-table |
| Load policy button | `load-policy-btn` | Load Q-table |
| Convergence indicator | `convergence-indicator` | Shows if converged |
| Learning status | `learning-status` | Training/idle status |

## In Scope
- Q-learning algorithm implementation
- Reward function design (game outcomes + user feedback)
- Policy updates (Q-table)
- Epsilon-greedy action selection
- Learning rate decay
- Model persistence (save/load Q-table)
- User feedback integration (thumbs up/down)
- Learning metrics dashboard
- Multi-game support (separate Q-tables)
- Convergence detection
- Visualization (win rate chart)

## Not In Scope
- Deep Q-learning (neural networks)
- Transfer learning across games
- Multi-agent learning
- Inverse reinforcement learning
- Policy gradient methods
- Advanced exploration strategies (UCB, Thompson sampling)

### Implementation
- [ ] Q-learning engine
- [ ] State representation for tic-tac-toe
- [ ] State representation for connect-four
- [ ] Action space definitions
- [ ] Reward functions
- [ ] Epsilon-greedy policy
- [ ] Learning rate scheduler
- [ ] Q-table persistence service
- [ ] User feedback handler
- [ ] Learning metrics tracker
- [ ] Convergence detector
- [ ] Dashboard UI components
- [ ] Win rate visualization

### Testing
- [ ] Unit tests for Q-learning algorithm
- [ ] Unit tests for reward calculation
- [ ] Unit tests for policy persistence
- [ ] Integration test: Learn tic-tac-toe (100 episodes)
- [ ] Integration test: User feedback updates policy
- [ ] Integration test: Policy loads after restart
- [ ] E2E test: `tests/journeys/reinforcement-learning.journey.spec.ts`
- [ ] Performance test: Training 1000 episodes <5 minutes

### Documentation
- [ ] Q-learning algorithm explained
- [ ] Reward function design rationale
- [ ] How to tune hyperparameters
- [ ] Adding new games to learning system

**Dependencies:** #9 (Tic-Tac-Toe), #34 (Connect Four), #36 (Maze Solve), #61 (Game Stats Tracking), #67 (Game Mode Visualizer)
**Blocks:** #87 (Predictive Behavior Engine)
**E2E Test:** `tests/journeys/reinforcement-learning.journey.spec.ts`
